vsock: Implement yaml configuration support

ToolmanP opened the pull request:
### Summary of the PR
Resolve #320 
- Prefer #[arg()] instead of #[clap()] in argument parsing suggested in [clap-rs v4](https://epage.github.io/blog/2022/09/clap4/#more-specific-derive-attributes)
- Use config-rs crate to implement yaml config parsing
- Write tests for new configuration file parsing

Proposed example config file:

#### config.yaml
```yaml
guest_cid: 4
socket: /tmp/vhost4.socket
uds-path: /tmp/vm4.vsock 
```
#### New Command

``` bash
user@localhost: ./vhost-user-vsock -f config.yaml
```

stefano-garzarella commented:
Thanks for the patches, following some suggestions to help reviewers and maintainability:
- use `vsock: ` prefix in each commit because this repository is a workspace with multiple crates
- add a description in each commit, for example explaining the motivation
- merge the fixes with the patches that introduced the problem, we usually prefer each patch to be buildable smoothly to simplify bisection

stefano-garzarella reviewed (changes requested):
Can we add in the `crates/vsock/README.md` the documentation of the configuration file?
I would add also an example in the `Usage` section,

What about using an array of objects (cid, socket, uds_path) in the YAML file?
For now we will support only 1 object (we can return an error if we have more than 1), but for the future we don't need to change its structure.

ToolmanP commented:
> Can we add the documentation of the configuration file in `crates/vsock/README.md`? I would add also an example in the `Usage` section.
> 
> What about using an array of objects (cid, socket, uds_path) in the YAML file? For now we will support only 1 object (we can return an error if we have more than 1), but for the future we don't need to change its structure.

Sure, i will get this work done and tweak the commit messages.

BTW, i'm thinking that the backend can start with no instance and works as a daemon to listen request via a socket.
Then we can write a ctl utility (vhostctl i guess?) to manage the instances through socket at runtime.

Plus, I have some diffculty when using clap-rs to support adding several instances in single command like:

``` console
user@localhost:~$ ./vhost-user-vsock
-- manual \
-- socket /tmp/vhost3.socket \
-- vds_path /tmp/vm3.vsock \
-- manual \
-- socket /tmp/vhost4.socket \
-- vds_path /tmp/vm4.vsock \
```
But It's kinda annoying that the clap-rs seems not to support this kind of operation. Maybe we should need another workaround or just use multiple configurations in a file instead.

stefano-garzarella commented:
> > Can we add the documentation of the configuration file in `crates/vsock/README.md`? I would add also an example in the `Usage` section.
> > What about using an array of objects (cid, socket, uds_path) in the YAML file? For now we will support only 1 object (we can return an error if we have more than 1), but for the future we don't need to change its structure.
> 
> Sure, i will get this work done and tweak the commit messages soon.

Yep ;-)

> 
> BTW, i'm thinking that the backend can start with no instance and works as a daemon to listen request via a socket. Then we can write a ctl utility (vhostctl i guess?) to manage the instances through socket at runtime.

What about if we send the new yaml through the socket, and we check the differences with the previous configuration?

> 
> Plus, I have some diffculty when using clap-rs to support adding several instances in single command like:
> 
> ```
> user@localhost:~$ ./vhost-user-vsock
> -- manual \
> -- socket /tmp/vhost3.socket \
> -- vds_path /tmp/vm3.vsock \
> -- manual \
> -- socket /tmp/vhost4.socket \
> -- vds_path /tmp/vm4.vsock \
> ```

Sorry, I don't get this example, can you elaborate a bit?

ToolmanP commented:
> > > Can we add the documentation of the configuration file in `crates/vsock/README.md`? I would add also an example in the `Usage` section.
> > > What about using an array of objects (cid, socket, uds_path) in the YAML file? For now we will support only 1 object (we can return an error if we have more than 1), but for the future we don't need to change its structure.
> > 
> > 
> > Sure, i will get this work done and tweak the commit messages soon.
> 
> Yep ;-)
> 
I've squashed the git messages into one and add vsock prefix to it. Is that OK?
Also I've updated the `Usage` in `crates/vsock/README.md`.
 
> > BTW, i'm thinking that the backend can start with no instance and works as a daemon to listen request via a socket. Then we can write a ctl utility (vhostctl i guess?) to manage the instances through socket at runtime.
> 
> What about if we send the new yaml through the socket, and we check the differences with the previous configuration?
> 
Yep, that's exactly I've planned!
> > Plus, I have some diffculty when using clap-rs to support adding several instances in single command like:
> > ```
> > user@localhost:~$ ./vhost-user-vsock
> > -- manual \
> > -- socket /tmp/vhost3.socket \
> > -- vds_path /tmp/vm3.vsock \
> > -- manual \
> > -- socket /tmp/vhost4.socket \
> > -- vds_path /tmp/vm4.vsock \
> > ```
> 
> Sorry, I don't get this example, can you elaborate a bit?
What I mean by that command is that we can create multiple vsockconfig via a single command
In particular,
`--spec` defines a vhost spec through command followed by the original (--guest_cid,--socket,--uds_path) 
to specify the VsockConfig.

Since this project aims to support multiple instances at the same time,
 I guess we may define the configuration procedure like this.

ToolmanP commented:
Also the CI check fails.

stefano-garzarella commented on crates/vsock/src/main.rs:
```diff
@@ -13,38 +13,83 @@ use std::{
+    spec: Option<VsockSpec>,
+
+    /// Load from a given configuration file
+    #[arg(short = 'f')]
```
please, use only long options like for the other options

ToolmanP replied:
This is now fixed.

stefano-garzarella commented:
> I've squashed the git messages into one and add vsock prefix to it. Is that OK? Also I've updated the `Usage` in `crates/vsock/README.md`.

No, please try to split in multiple commits and explain the reason of the changes in the commit description.
Like why we need Deserialize, and other things you explained in the PR description.

> 
> > > BTW, i'm thinking that the backend can start with no instance and works as a daemon to listen request via a socket. Then we can write a ctl utility (vhostctl i guess?) to manage the instances through socket at runtime.
> > 
> > 
> > What about if we send the new yaml through the socket, and we check the differences with the previous configuration?
> 
> Yep, that's exactly I've planned!
> 
> > > Plus, I have some diffculty when using clap-rs to support adding several instances in single command like:
> > > ```
> > > user@localhost:~$ ./vhost-user-vsock
> > > -- manual \
> > > -- socket /tmp/vhost3.socket \
> > > -- vds_path /tmp/vm3.vsock \
> > > -- manual \
> > > -- socket /tmp/vhost4.socket \
> > > -- vds_path /tmp/vm4.vsock \
> > > ```
> > 
> > 
> > Sorry, I don't get this example, can you elaborate a bit?
> 
> What I mean by that command is that we can create multiple vsockconfig via a single command In particular, `--spec` defines a vhost spec through command followed by the original (--guest_cid,--socket,--uds_path) to specify the VsockConfig.

Sorry, I'm lost, what `--spec` is? what `--manual` is?
In the example, are you referring to `./vhost-user-vsock` or to a new tool to generate the configuration?

> 
> Since this project aims to support multiple instances at the same time, I guess we may define the configuration procedure like this.

How is this related to updating the configuration file as we said before?

stefano-garzarella commented on crates/vsock/README.md:
````diff
@@ -43,6 +43,16 @@ Run the vhost-user-vsock device:
+Configuration Example
+
+```yaml
+vhost:
````
What `vhost` means here?

ToolmanP replied:
I originally thought  about spliting host part from the sibling part.  Was that sensible? Now I switched to your suggested

``` yaml
vms:
...
```

stefano-garzarella commented on crates/vsock/src/main.rs:
```diff
@@ -13,38 +13,83 @@ use std::{
+                {
+                    Ok(s) => {
+                        let mut specs = s.get::<Vec<VsockSpec>>("vhost").unwrap();
+                        specs.pop().map(|spec| {
```
Should we return an error if `specs.len()` is `!= 1`?

What `specs` means? (I would use the same name of the YAML array, maybe `vms` or `vsock-devices`

ToolmanP replied:
This is resolved now.

stefano-garzarella commented on crates/vsock/src/main.rs:
```diff
@@ -13,38 +13,83 @@ use std::{
-#[clap(version, about, long_about = None)]
-struct VsockArgs {
+#[derive(Args, Debug, Deserialize)]
+struct VsockSpec {
```
What about `VsockParam`?

ToolmanP replied:
Yep I fixed it.

ToolmanP commented on crates/vsock/src/main.rs:
```diff
@@ -13,38 +13,83 @@ use std::{
+use serde::Deserialize;
+use vhost::{
+    vhost_user,
+    vhost_user::{message::VhostUserConfig, Listener},
```
My bad this was  merely a typo and rust-analyzer introduced this field to this file automatically. I forgot to remove this.

ToolmanP commented:
> > I've squashed the git messages into one and add vsock prefix to it. Is that OK? Also I've updated the `Usage` in `crates/vsock/README.md`.
> 
> No, please try to split in multiple commits and explain the reason of the changes in the commit description. Like why we need Deserialize, and other things you explained in the PR description.

I re-split the commit into multiple commits and wrote the description according to each change. Because I used to mistake the title for the description for a long time. Sorry for my incompetence. They should be elaborate enough to know the motivation behind and changes I made for each commit.
> 
> > > > BTW, i'm thinking that the backend can start with no instance and works as a daemon to listen request via a socket. Then we can write a ctl utility (vhostctl i guess?) to manage the instances through socket at runtime.
> > > 
> > > 
> > > What about if we send the new yaml through the socket, and we check the differences with the previous configuration?
> > 
> > 
> > Yep, that's exactly I've planned!
> > > > Plus, I have some diffculty when using clap-rs to support adding several instances in single command like:
> > > > ```
> > > > user@localhost:~$ ./vhost-user-vsock
> > > > -- manual \
> > > > -- socket /tmp/vhost3.socket \
> > > > -- vds_path /tmp/vm3.vsock \
> > > > -- manual \
> > > > -- socket /tmp/vhost4.socket \
> > > > -- vds_path /tmp/vm4.vsock \
> > > > ```
> > > 
> > > 
> > > Sorry, I don't get this example, can you elaborate a bit?
> > 
> > 
> > What I mean by that command is that we can create multiple vsockconfig via a single command In particular, `--spec` defines a vhost spec through command followed by the original (--guest_cid,--socket,--uds_path) to specify the VsockConfig.
> 
> Sorry, I'm lost, what `--spec` is? what `--manual` is? In the example, are you referring to `./vhost-user-vsock` or to a new tool to generate the configuration?
This is merely a typo by me. 
> 
> > Since this project aims to support multiple instances at the same time, I guess we may define the configuration procedure like this.
> 
> How is this related to updating the configuration file as we said before?
Really sorry that I may not convey this idea clearly enough because I'm not a native speaker.

Since the configuration can accept an array of vms, like

``` yaml
vms:
 - guest_cid: 3
   socket: /tmp/vm3.socket
   uds_path: /tmp/vm3.sock
 - guest_cid: 4
   socket: /tmp/vm4.socket
   uds_path:/tmp/vm4.socket
```
and parse them into `VsockParam` one by one, 
Then I think we might be able to enable users to add multiple vms from the command line with something like
``` console
user@localhost:~$ vhost-user-vsock --param
--guest_cid 3
--socket /tmp/vm3.socket
--uds_path/tmp/vm3.sock
--param
--guest_cid 4
--socket /tmp/vm4.socket
--uds_path/tmp/vm4.sock
```
I think that makes the config from cmd  more consistent with config from file. :)

stefano-garzarella commented on crates/vsock/src/main.rs:
```diff
@@ -40,10 +40,10 @@ struct VsockSpec {
 
     /// Load from a given configuration file
-    #[arg(short = 'f')]
+    #[arg(long)]
```
Also for this

stefano-garzarella commented:
> I re-split the commit into multiple commits and wrote the description according to each change. Because I used to mistake the title for the description for a long time. Sorry for my incompetence. They should be elaborate enough to know the motivation behind and changes I made for each commit.

It is better, but I suggest you make the changes as soon as possible in the patches. (e.g. `VsockSpec` can already be `VsockParam` when you introduce it).

> 
> Really sorry that I may not convey this idea clearly enough because I'm not a native speaker.

Not a problem at all! I'm not a native speaker too ;-)

> 
> Since the configuration can accept an array of vms, like
> 
> ```yaml
> vms:
>  - guest_cid: 3
>    socket: /tmp/vhost3.socket
>    uds_path: /tmp/vm3.sock
>  - guest_cid: 4
>    socket: /tmp/vhost4.socket
>    uds_path: /tmp/vm4.socket
> ```
> 
> and parse them into `VsockParam` one by one, Then I think we might be able to enable users to add multiple vms from the command line with something like
> 
> ```
> user@localhost:~$ vhost-user-vsock
> --param
> --guest_cid 3
> --socket /tmp/vhost3.socket
> --uds_path/tmp/vm3.sock
> --param
> --guest_cid 4
> --socket /tmp/vhost4.socket
> --uds_path/tmp/vm4.sock
> ```
> 
> I think that will make the config from cmd more consistent with config from file. :)

Right, I agree. I'm not sure if we need `--param`, but we need to think at it better when we will do this change.
Thanks for more details, it is now clear to me and it makes sense!

ToolmanP commented on crates/vsock/src/main.rs:
```diff
@@ -40,10 +40,10 @@ struct VsockSpec {
 struct VsockArgs {
     #[command(flatten)]
-    spec: Option<VsockSpec>,
+    param: Option<VsockParam>,
```
Sure. I gotta spend some time to rebase this.

ToolmanP replied:
It's fixed now. More consistent commits.

ToolmanP commented on crates/vsock/src/main.rs:
```diff
@@ -105,20 +142,22 @@ mod tests {
-        fn from_args(guest_cid: u64, socket: &str, uds_path: &str) -> Self {
-            VsockArgs {
+    impl VsockArgs { 
+        fn from_args(guest_cid: u64, socket: &str, uds_path: &str, daemonize: bool, config: &str) -> Self {
```
Originally, I planned to daemonize this backend from the first place. But afterwards, I changed my idea to finish the configuration work first. However, I forgot to clean up this line of code.  This is merely a really bad habit of me doing commits. Give me some time and I will fix this.

stefano-garzarella replied:
Sure, don't worry, it is a way to learn the process!

stefano-garzarella commented:
I forgot to mention that usually we put `Signed-off-by:` at the end of the commit description, like a sign.

ToolmanP commented:
> @ToolmanP in `main` now there is a new `rust-vmm-ci` checkout, so please rebase this PR on the current `main` and the CI should be fixed.

I've rebased to the upstream but something weird happens in the pipeline with the error

``` console
$ /etc/buildkite-agent/hooks/pre-exit
docker: error during connect: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create": EOF.
```

I tried the same command on my local machines. There are no warnings indeed.

ToolmanP commented:
> I forgot to mention that usually we put `Signed-off-by:` at the end of the commit description, like a sign.

Along with previous ones, I've cleaned up the code in previous commits  and reformat the messages. please check :-)

stefano-garzarella commented:
> > @ToolmanP in `main` now there is a new `rust-vmm-ci` checkout, so please rebase this PR on the current `main` and the CI should be fixed.
> 
> I've rebased to the upstream but something weird happens in the pipeline with the error
> 
> ```
> $ /etc/buildkite-agent/hooks/pre-exit
> docker: error during connect: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create": EOF.
> ```
> 
> I tried the same command on my local machines. There are no warnings indeed.

Yep, I think there is something not working in our CI

ToolmanP commented:
> There are still some things not cleaned up for example (vds_path in the first commit, becomes uds_path in the second), maybe it's better to squash everything in a single commit as the code is now.
> 
> The final result LGTM.

Everything is now squashed into a single commit.

vireshk commented on crates/vsock/src/main.rs:
```diff
@@ -10,38 +10,81 @@ mod vsock_conn;
+                }
+            }
+            _ => None,
+        }
```
What about using "if let" for both the match blocks ?

ToolmanP replied:
```
impl VsockArgs {
    pub fn parse_config(&self) -> Option<VsockConfig> {
        if let Some(s) = &self.config {
            let b = config::Config::builder()
                .add_source(config::File::new(s.as_str(), config::FileFormat::Yaml))
                .build();
            if let Ok(c) = b {
                let mut v = c.get::<Vec<VsockParam>>("vms").unwrap();
                if v.len() != 1 {
                    return None;
                }
                v.pop()
                    .map(|vm| VsockConfig::new(vm.guest_cid, vm.socket, vm.uds_path))
            } else {
                None
            }
        } else {
            None
        }
    }
}
```

You mean something like this?

vireshk replied:
pub fn parse_config(&self) -> Option<VsockConfig> {
        if let Some(c) = &self.config {
                if let Ok(s) = config::Config::builder()
                    .add_source(config::File::new(c.as_str(), config::FileFormat::Yaml))
                    .build()
                {
                        let mut v = s.get::<Vec<VsockParam>>("vms").unwrap();
                        if v.len() == 1 {
                            return v.pop()
                                .map(|vm| VsockConfig::new(vm.guest_cid, vm.socket, vm.uds_path));
                        }
                }
            }
        }

	None
    }

vireshk replied:
@ToolmanP I can see that you have resolved this, but there were few more suggestions in above code block to reduce code size and avoid the extra return None. Won't it be worth doing ?

stefano-garzarella replied:
@ToolmanP IIUC @vireshk is referring to:
```
            if let Ok(s) = config::Config::builder()
                .add_source(config::File::new(c.as_str(), config::FileFormat::Yaml))
                .build()
            {
```

Fixed that, I think we are ready to merge this PR :-)

vireshk commented on crates/vsock/src/main.rs:
```diff
@@ -10,38 +10,81 @@ mod vsock_conn;
-        Ok(VsockConfig::new(cmd_args.guest_cid, socket, uds_path))
+        // we try to use the configuration first, if failed,  then fall back to the manual settings.
+        let config: Option<Self> = cmd_args.parse_config();
+        if let Some(c) = config {
```
While at this place I belive a match may look better, since we want to write code for all match cases.

vireshk replied:
No, but something like this:

        match config {
            Some(c) => Ok(c),
	    _ => {
              cmd_args.param.map_or(Err(Error::ConfigParse), |p| {
                  Ok(Self::new(
                      p.guest_cid,
                      p.socket.trim().to_string(),
                      p.uds_path.trim().to_string(),
                  ))
              })
	    }
       }

vireshk commented on crates/vsock/src/main.rs:
```diff
@@ -90,7 +133,6 @@ pub(crate) fn start_backend_server(config: VsockConfig) {
 
 fn main() {
     env_logger::init();
-
```
I would leave the blank line as is.

ToolmanP replied:
Alright I will tweak this.

ToolmanP replied:
It's tweaked now.

ToolmanP commented:
I've modified the configuration example and some semantics based on #319. Please review those stuff. thx :-)

ToolmanP commented:
It's cleared now.

stefano-garzarella commented:
@ToolmanP thanks for rebasing!

LGTM, just fix the @vireshk suggestion and we can merge this!

techiepriyansh commented:
> > Since the configuration can accept an array of vms, like
> > ```yaml
> > vms:
> >  - guest_cid: 3
> >    socket: /tmp/vhost3.socket
> >    uds_path: /tmp/vm3.sock
> >  - guest_cid: 4
> >    socket: /tmp/vhost4.socket
> >    uds_path: /tmp/vm4.socket
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and parse them into `VsockParam` one by one, Then I think we might be able to enable users to add multiple vms from the command line with something like
> > ```
> > user@localhost:~$ vhost-user-vsock
> > --param
> > --guest_cid 3
> > --socket /tmp/vhost3.socket
> > --uds_path/tmp/vm3.sock
> > --param
> > --guest_cid 4
> > --socket /tmp/vhost4.socket
> > --uds_path/tmp/vm4.sock
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think that will make the config from cmd more consistent with config from file. :)
> 
> Right, I agree. I'm not sure if we need `--param`, but we need to think at it better when we will do this change. Thanks for more details, it is now clear to me and it makes sense!

Now that we are starting to support multiple guests, we need to decide up on this.

stefano-garzarella commented:
@techiepriyansh yep, I think so.

What is the easiest way to implement it?

I was thinking something like:
```
vhost-user-vsock --vm 3,/tmp/vhost3.socket,/tmp/vm3.sock
```

But I don't know if it's very usable. Anyway, we could also force the user to use YAML for a configuration with multiple VMs.

techiepriyansh commented:
> I was thinking something like:
> 
> ```
> vhost-user-vsock --vm 3,/tmp/vhost3.socket,/tmp/vm3.sock
> ```

This looks good. Maybe we can make it more explicit by using something like:
```
vhost-user-vsock --vm guest_cid=3,socket=/tmp/vhost3.socket,uds_path=/tmp/vm3.sock
```

> But I don't know if it's very usable. Anyway, we could also force the user to use YAML for a configuration with multiple VMs.

Agreed. I think I'll go ahead with this for now. We can come back to extending the CLI interface later if needed.
